### PR TITLE
[Fusion] [Graph]Add Matmul Allreduce Rmsnorm fusion Pass

### PR DESCRIPTION
**Pull request by wxsIcey**
This PR add `MatmulAllreduceRmsnorm` operator and introduces a graph fusion pass for `matmul_allreduce_rmsnorm` operations. The implementation includes a new configuration flag, a pattern matching pass using `torch._inductor.pattern_matcher`.

Co-authored-by: Trunrain [270250579@qq.com](mailto:270250579@qq.com)

- vLLM version: v0.12.0
- vLLM main: https://github.com/vllm-project/vllm/commit/ad32e3e19ccf0526cb6744a5fed09a138a5fb2f9

### CODE REVIEW

File: vllm_ascend/worker/model_runner_v1.py [hunk 1]
```diff
@@ -1135,6 +1141,12 @@ def _generate_process_reqs_hidden_states(self, maybe_padded_num_tokens,
             pad_size = get_forward_context().pad_size
             if pad_size > 0:
                 hidden_states = hidden_states[:-pad_size, :]
+        ascend_compilation_config: dict = config.additional_config.get(
+            "ascend_compilation_config", {})
+        if ascend_compilation_config.get("fuse_allreduce_rms", True) and not isinstance(
```
Comment (realliujiaxu) [inline vllm_ascend/worker/model_runner_v1.py]: there is no code to run under this if statement

File: vllm_ascend/worker/model_runner_v1.py [hunk 2]
```diff
@@ -1096,6 +1096,12 @@ def _generate_process_reqs_hidden_states(self, maybe_padded_num_tokens,
                                              input_ids, positions,
                                              intermediate_tensors,
                                              inputs_embeds):
+        device_group = get_tp_group().device_group
+        backend = device_group._get_backend(torch.device("npu"))
+        local_rank = torch.distributed.get_rank(group=device_group)
+        tp_group_name = backend.get_hccl_comm_name(ocal_rank)
```
Comment (realliujiaxu) [inline vllm_ascend/worker/model_runner_v1.py]: is `ocal_rank` a typo?

File: vllm_ascend/patch/worker/patch_unquantized_gemm.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+def default_unquantized_gemm(
+    layer: torch.nn.Module,
+    x: torch.Tensor,
+    weight: torch.Tensor,
+    bias: torch.Tensor | None = None,
+) -> torch.Tensor:
+    return torch.ops.vllm.unquantized_gemm(x, weight, bias)
+
+
+vllm.model_executor.layers.utils.default_unquantized_gemm = default_unquantized_gemm
```
Comment (weijinqian0) [inline vllm_ascend/patch/worker/patch_unquantized_gemm.py:57]: need check why need this pacth.

File: vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+from torch._inductor.pattern_matcher import (PatternMatcherPass,
+                                             PatternPrettyPrinter)
+from vllm.compilation.vllm_inductor_pass import VllmInductorPass
+from vllm.config import VllmConfig
+from vllm.distributed import tensor_model_parallel_all_reduce, get_tensor_model_parallel_world_size
+from vllm.distributed.parallel_state import get_tp_group
+from vllm.logger import logger
+
+
+class MatmulAllReduceAddRMSNormPattern:
```
Comment (wxsIcey) [inline vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py]: rename pass name

File: vllm_ascend/patch/worker/patch_distributed.py [hunk 5]
```diff
@@ -111,5 +111,8 @@ def all_to_all(self,
                                                    gather_dim, scatter_sizes,
                                                    gather_sizes)
 
+    def all_reduce(self, input_):
```
Comment (wxsIcey) [inline vllm_ascend/patch/worker/patch_distributed.py:115]: change patch init file

File: vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        rms_norm_weight = torch.randn(hidden_size, device="npu")
+        return [x, weight, residual, rms_norm_weight]
+
+    def register(self, pm_pass: PatternMatcherPass):
+
+        def pattern(x, weight, residual, rms_norm_weight):
+            mm = torch.ops.vllm.unquantized_gemm(x, weight, None)
+            all_reduce_ = tensor_model_parallel_all_reduce(mm)
+            output = torch.ops.npu.npu_add_rms_norm.default(
+                all_reduce_, residual, rms_norm_weight)
```
Comment (wxsIcey) [inline vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py]: remove .default

File: vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        pm.register_replacement(pattern, replacement, self.get_inputs(),
+                                pm.fwd_only, pm_pass)
+
+class MatmulAllReduceAddRMSNormPass(VllmInductorPass):
+
+    def __init__(self, vllm_config: VllmConfig):
+        super().__init__(vllm_config)
+        self.pattern_match_passes: PatternMatcherPass = PatternMatcherPass(
+            pass_name="allreduce_rmsnorm_fusion_pass")
+
```
Comment (wxsIcey) [inline vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py:132]: pass name should change?

File: vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        pattern_idx = 0
+        for pattern_entry in self.pattern_match_passes.patterns.values():
+            for p in pattern_entry:
+                p_str = PatternPrettyPrinter.run(p.pattern)
+                logger.debug("Pattern %d: %s", pattern_idx, p_str)
+                pattern_idx += 1
+        logger.debug("Replaced %s patterns", self.matched_count)
+        self.end_and_log()
+
+    def is_applicable(self, runtime_shape):
```
Comment (wxsIcey) [inline vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py]: change to is_applicable_for_range

File: vllm_ascend/patch/worker/__init__.py [hunk 9]
```diff
@@ -21,10 +21,13 @@
     import vllm_ascend.patch.worker.patch_triton
 
 # isort: off
+import vllm_ascend.ops.register_custom_ops  # noqa
```
Comment (realliujiaxu) [inline vllm_ascend/patch/worker/__init__.py]: what is this for?

File: vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+            out0, out1 = torch.ops._C_ascend.matmul_allreduce_add_rmsnorm(
+                x, weight, residual, rms_norm_weight, self.tp_group_name,
+                self.tp_size, self.local_rank, self.eps, True,
+                False)
+            return out0, out1
+
+        pm.register_replacement(pattern, replacement, self.get_inputs(),
+                                pm.fwd_only, pm_pass)
+
+class MatmulAllreduceAddRMSNormAllgather:
```
Comment (realliujiaxu) [inline vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py]: change it to `LastLayerMatmulAllreduceAddRMSNormAllgatherPattern`

File: vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py (hunk 7 above)
Comment (realliujiaxu) [reply vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py:132]: I suggest leave pass  and pattern name as `MatmulAllReduceAddRMSNormPass` and `xxMatmulAllReduceAddRMSNormPattern` respectively. And using comments to explain that the fusion operator actually splits the allreduce into reducescatter and allgather.

File: vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+from torch._inductor.pattern_matcher import (PatternMatcherPass,
+                                             PatternPrettyPrinter)
+from vllm.compilation.vllm_inductor_pass import VllmInductorPass
+from vllm.config import VllmConfig
+from vllm.distributed import tensor_model_parallel_all_reduce, get_tensor_model_parallel_world_size
+from vllm.distributed.parallel_state import get_tp_group
+from vllm.logger import logger
+
+
+class MatmulReduceScatterAddRMSNormAllgatherPattern:
```
Comment (realliujiaxu) [inline vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py]: change it to `MiddleLayerXXX`

**Review by realliujiaxu — CHANGES_REQUESTED** [review]
What model was used for the verification? It would be great if you could add the performance data. This whill allow users to assess whether the performance benefit meet their expectations during usage.

File: vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py (hunk 7 above)
Comment (wxsIcey) [reply vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py:132]: Thanks. I have modified it.

File: vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py (hunk 10 above)
Comment (wxsIcey) [reply vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py]: done

File: vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py (hunk 11 above)
Comment (wxsIcey) [reply vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py]: done

File: vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+                logger.debug("Pattern %d: %s", pattern_idx, p_str)
+                pattern_idx += 1
+        logger.debug("Replaced %s patterns", self.matched_count)
+        self.end_and_log()
+
+    def is_applicable_for_range(self, compile_range: Range) -> bool:
+        """
+        Check if the pass is applicable for the current configuration.
+        """
+        return True
```
Comment (realliujiaxu) [inline vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py]: Does this pass provide performance benefits for any compile_range? If not, this pass should be turned on depends on specific compile_range.

**Comment by Trunrain** [comment]
> What model was used for the verification? It would be great if you could add the performance data. This whill allow users to assess whether the performance benefit meet their expectations during usage.

Verified in the 8TP scenario on Qwen3 32B dense model(hardware 910b4).
| Metric | fuse_allreduce_rms | SP |
|--------|--------------------|----------------|
| Model & Setup | qwen3 32B 8TP (input_len: 10k, output_len: 2k) | qwen3 32B 8TP (input_len: 10k, output_len: 2k) |
| `VLLM_ASCEND_ENABLE_NZ` | 0 | 1 |
| `VLLM_ASCEND_ENABLE_DENSE_OPTIMIZE` | 0 | 1 |
| `VLLM_ASCEND_ENABLE_FLASHCOMM` | 0 | 1 |
| **Median TTFT (ms)** | 177635.79 | 184819.24 |

File: vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py (hunk 12 above)
Comment (Trunrain) [reply vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py]: The current computation-communication tiling strategy is set to 512[input_len]. It is recommended to enable it for values greater than 512.

File: vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+                pattern_idx += 1
+        logger.debug("Replaced %s patterns", self.matched_count)
+        self.end_and_log()
+
+    def is_applicable_for_range(self, compile_range: Range) -> bool:
+        """
+        Check if the pass is applicable for the current configuration.
+        """
+        # computation-communication tiling block is 512
+        applicable = compile_range.start > 512
```
Comment (realliujiaxu) [inline vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py]: `512` should be added into `compile_ranges_split_points `, otherwise `is_applicable_for_range` always return `False`. Please refer to midification of `vllm/config/vllm.py` in https://github.com/vllm-project/vllm/pull/24252:
```python
        if compilation_config.pass_config.fuse_allreduce_rms:
            tp_size = self.parallel_config.tensor_parallel_size
            max_size = compilation_config.pass_config.flashinfer_max_size(tp_size)
            if max_size is not None:
                max_token_num = max_size // (
                    self.model_config.get_hidden_size()
                    * self.model_config.dtype.itemsize
                )
                if (
                    max_num_batched_tokens is not None
                    and max_token_num < max_num_batched_tokens
                ):
                    computed_compile_ranges_split_points.append(max_token_num)
```

Comment (wxsIcey) [reply vllm_ascend/compilation/passes/allreduce_rmsnorm_fusion_pass.py]: Yes, the current approach has the problem. I will change the implement, you cantake a look at the logic in this section. @Trunrain
